File: libs/edis-ts-contracts/src/findings.ts
/**
 * Zod mirror of `edis_contracts/findings.py` (intelligence-layer contracts).
 *
 * Every figure on a `Finding` is computed by a detector; the LLM never overrides
 * it. `narrative` is null until grounded narration succeeds.
 */
import { z } from "zod";
import { datetime, schemaVersion, strMap, uuid } from "./common.js";

/** Mirror of `FindingKind(str, Enum)`. */
export const FindingKindSchema = z.enum([
  "point_anomaly",
  "seasonal_break",
  "level_shift",
  "trend_break",
  "forecast_breach",
  "root_cause",
]);
export type FindingKind = z.infer<typeof FindingKindSchema>;

/** Mirror of `CandidateCause` (lag-adjusted RCA correlate). */
export const CandidateCauseSchema = z.object({
  metric_key: z.string(),
  dimensions: strMap().default({}),
  correlation: z.number(),
  lag_minutes: z.number().int(),
  contribution_pct: z.number().nullish(),
  direction: z.enum(["leading", "coincident", "lagging"]),
  observed_delta: z.number(),
});
export type CandidateCause = z.infer<typeof CandidateCauseSchema>;

/** Mirror of `EvidenceItem` (one computed fact the narrator may cite). */
export const EvidenceItemSchema = z.object({
  kind: z.string(),
  metric_key: z.string().nullish(),
  dimensions: strMap().default({}),
  summary: z.string(),
  values: z.record(z.string(), z.number()).default({}),
  ref: z.record(z.string(), z.unknown()).nullish(),
});
export type EvidenceItem = z.infer<typeof EvidenceItemSchema>;

/** Mirror of `EvidenceBundle` -- the only thing the LLM may reason over. */
export const EvidenceBundleSchema = z.object({
  bundle_id: uuid(),
  tenant_id: z.string(),
  finding_id: uuid(),
  created_at: datetime(),
  items: z.array(EvidenceItemSchema).default([]),
  allowed_numbers: z.array(z.number()).default([]),
  schema_version: schemaVersion(),
});
export type EvidenceBundle = z.infer<typeof EvidenceBundleSchema>;

/** Mirror of `Finding` -- payload of `edis.findings.v1`. */
export const FindingSchema = z.object({
  finding_id: uuid(),
  tenant_id: z.string(),
  kind: FindingKindSchema,
  metric_key: z.string(),
  dimensions: strMap().default({}),
  window_start: datetime(),
  window_end: datetime(),
  detector: z.string(),
  detector_version: z.string(),
  observed_value: z.number(),
  expected_value: z.number(),
  deviation: z.number(),
  deviation_pct: z.number(),
  score: z.number(),
  severity: z.number(),
  confidence: z.number(),
  business_impact_input: z.number(),
  candidate_causes: z.array(CandidateCauseSchema).default([]),
  narrative: z.string().nullish(),
  narrative_model: z.string().nullish(),
  evidence_ref: uuid().nullish(),
  status: z
    .enum(["open", "acknowledged", "resolved", "expired"])
    .default("open"),
  created_at: datetime(),
  schema_version: schemaVersion(),
});
export type Finding = z.infer<typeof FindingSchema>;

/** Mirror of `Forecast` -- payload of `edis.forecasts.v1`. */
export const ForecastSchema = z.object({
  forecast_id: uuid(),
  tenant_id: z.string(),
  metric_key: z.string(),
  dimensions: strMap().default({}),
  model: z.string(),
  horizon_days: z.number().int(),
  points: z.array(z.record(z.string(), z.unknown())).default([]),
  generated_at: datetime(),
  schema_version: schemaVersion(),
});
export type Forecast = z.infer<typeof ForecastSchema>;

File: libs/edis-contracts/edis_contracts/findings.py
"""Intelligence-layer contracts: findings, candidate causes, evidence, forecasts.

The LLM never invents numbers. Every figure on a :class:`Finding` is computed by a
detector; the narrator may only reason over the :class:`EvidenceBundle`, whose
``allowed_numbers`` whitelist the grounding guard enforces.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


class FindingKind(str, Enum):
    POINT_ANOMALY = "point_anomaly"
    SEASONAL_BREAK = "seasonal_break"
    LEVEL_SHIFT = "level_shift"
    TREND_BREAK = "trend_break"
    FORECAST_BREACH = "forecast_breach"
    ROOT_CAUSE = "root_cause"


class CandidateCause(BaseModel):
    """A lag-adjusted, ranked correlate of an anomaly (RCA output)."""

    metric_key: str
    dimensions: dict[str, str] = Field(default_factory=dict)
    correlation: float  # -1..1, lag-adjusted
    lag_minutes: int
    contribution_pct: float | None = None
    direction: Literal["leading", "coincident", "lagging"]
    observed_delta: float


class EvidenceItem(BaseModel):
    """One computed fact the narrator is permitted to cite."""

    kind: str  # "metric_window" | "candidate_cause" | "baseline" | "forecast"
    metric_key: str | None = None
    dimensions: dict[str, str] = Field(default_factory=dict)
    summary: str
    values: dict[str, float] = Field(default_factory=dict)
    ref: dict | None = None  # pointer to the source rows


class EvidenceBundle(BaseModel):
    """The *only* thing the LLM may reason over when narrating a finding."""

    bundle_id: UUID
    tenant_id: str
    finding_id: UUID
    created_at: datetime
    items: list[EvidenceItem] = Field(default_factory=list)
    allowed_numbers: list[float] = Field(default_factory=list)  # grounding-guard whitelist
    schema_version: int = 1


class Finding(BaseModel):
    """Payload of ``edis.findings.v1`` -- an atomic, computed detection.

    ``observed_value``/``expected_value``/``deviation`` are computed and the LLM
    never overrides them; ``narrative`` is null until grounded narration succeeds.
    """

    finding_id: UUID
    tenant_id: str
    kind: FindingKind
    metric_key: str
    dimensions: dict[str, str] = Field(default_factory=dict)
    window_start: datetime
    window_end: datetime
    detector: str
    detector_version: str
    observed_value: float
    expected_value: float
    deviation: float
    deviation_pct: float
    score: float  # detector-native (z-score / residual)
    severity: float  # 0..1 normalized
    confidence: float  # 0..1
    business_impact_input: float  # 0..1 -- Decision owns final ranking
    candidate_causes: list[CandidateCause] = Field(default_factory=list)
    narrative: str | None = None  # grounded LLM text; null until narrated
    narrative_model: str | None = None  # e.g. "claude-opus-4-8"
    evidence_ref: UUID | None = None  # FK to persisted EvidenceBundle (audit)
    status: Literal["open", "acknowledged", "resolved", "expired"] = "open"
    created_at: datetime
    schema_version: int = 1


class Forecast(BaseModel):
    """Payload of ``edis.forecasts.v1`` -- point + prediction interval per metric."""

    forecast_id: UUID
    tenant_id: str
    metric_key: str
    dimensions: dict[str, str] = Field(default_factory=dict)
    model: str  # "statsforecast.AutoETS"
    horizon_days: int
    points: list[dict] = Field(default_factory=list)  # [{ts, yhat, yhat_lower, yhat_upper}]
    generated_at: datetime
    schema_version: int = 1
